chronos/threadsync.nim

```diff
@@ -0,0 +1,344 @@
+      if not(wres.get()):
+        return ok(false)
+      let res = handleEintr(read(eventFd, addr data, sizeof(uint64)))
+      if res < 0:
+        let errorCode = osLastError()
+        # If errorCode == EAGAIN it means that reading operation is already
+        # pending and so some other consumer reading eventfd or pipe end, in
+        # this case we going to ignore error and wait for another event.
+        if errorCode != EAGAIN:
+          return err(osErrorMsg(errorCode))
+      elif res != sizeof(data):
+        return err(osErrorMsg(EINVAL))
+      else:
+        return ok(true)
+
+proc fire*(signal: ThreadSignalPtr): Future[void] =
+  ## Set state of ``signal`` to signaled in asynchronous way.
+  var retFuture = newFuture[void]("asyncthreadsignal.fire")
+  when defined(windows):
+    if setEvent(signal[].event) == 0'u32:
+      retFuture.fail(newException(AsyncError, osErrorMsg(osLastError())))
+    else:
+      retFuture.complete()
+  else:
+    var data = 1'u64
+    let eventFd =
+      when defined(linux):
+        cint(signal[].efd)
+      else:
+        cint(signal[].wfd)
+
+    proc continuation(udata: pointer) {.gcsafe, raises: [].} =
+      if not(retFuture.finished()):
+        let res = handleEintr(write(eventFd, addr data, sizeof(uint64)))
+        if res < 0:
+          let errorCode = osLastError()
+          discard removeWriter2(AsyncFD(eventFd))
+          retFuture.fail(newException(AsyncError, osErrorMsg(errorCode)))
+        elif res != sizeof(data):
+          discard removeWriter2(AsyncFD(eventFd))
+          retFuture.fail(newException(AsyncError, osErrorMsg(EINVAL)))
+        else:
+          let eres = removeWriter2(AsyncFD(eventFd))
+          if eres.isErr():
+            retFuture.fail(newException(AsyncError, osErrorMsg(eres.error)))
+          else:
+            retFuture.complete()
+
+    proc cancellation(udata: pointer) {.gcsafe, raises: [].} =
+      if not(retFuture.finished()):
+        discard removeWriter2(AsyncFD(eventFd))
+
+    let res = handleEintr(write(eventFd, addr data, sizeof(uint64)))
+    if res < 0:
+      let errorCode = osLastError()
+      case errorCode
+      of EAGAIN:
+        let loop = getThreadDispatcher()
+        if not(loop.contains(AsyncFD(eventFd))):
+          let rres = register2(AsyncFD(eventFd))
+          if rres.isErr():
+            retFuture.fail(newException(AsyncError, osErrorMsg(rres.error)))
+            return retFuture
+        let wres = addWriter2(AsyncFD(eventFd), continuation)
+        if wres.isErr():
+          retFuture.fail(newException(AsyncError, osErrorMsg(wres.error)))
+        else:
+          retFuture.cancelCallback = cancellation
+      else:
+        retFuture.fail(newException(AsyncError, osErrorMsg(errorCode)))
+    elif res != sizeof(data):
+      retFuture.fail(newException(AsyncError, osErrorMsg(EINVAL)))
+    else:
+      retFuture.complete()
+
+  retFuture
+
+when defined(windows):
+  proc wait*(signal: ThreadSignalPtr) {.async.} =
```

Comment:
hm, what are the calling constraints here for `fire` and `wait`? ie which threads can they each be called from?

what happens for example if I `wait` from multiple threads?

Reply:
So for any number callers to `fire()/fireSync()` from single or many threads and for any number of waiters via `wait()/waitSync()` for single or many thread, every call to `fire()/fireSync()` will wake up only one waiter via `wait()/waitSync()`, but ThreadSignal is not a semaphore so there could be only one pending `fire()` signal until waiter will not receive it.